core/chains/evm/txmgr/finalizer.go

```diff
@@ -280,16 +348,290 @@ func (f *evmFinalizer) batchCheckReceiptHashesOnchain(ctx context.Context, block
+		} else {
+			output = data.Receipt
+		}
+
+		f.lggr.Debugw("Callback: resuming tx with receipt", "output", output, "taskErr", taskErr, "pipelineTaskRunID", data.ID)
+		if err := f.resumeCallback(ctx, data.ID, output, taskErr); err != nil {
+			return fmt.Errorf("failed to resume suspended pipeline run: %w", err)
+		}
+		// Mark tx as having completed callback
+		if err := f.txStore.UpdateTxCallbackCompleted(ctx, data.ID, f.chainID); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *evmFinalizer) ProcessOldTxsWithoutReceipts(ctx context.Context, oldTxIDs []int64, head, latestFinalizedHead *evmtypes.Head) error {
+	if len(oldTxIDs) == 0 {
+		return nil
+	}
+	oldTxs, err := f.txStore.FindTxesByIDs(ctx, oldTxIDs, f.chainID)
+	if err != nil {
+		return fmt.Errorf("failed to find transactions with IDs: %w", err)
+	}
+
+	var errorList []error
+	for _, oldTx := range oldTxs {
+		f.lggr.Criticalw(fmt.Sprintf("transaction with ID %v expired without ever getting a receipt for any of our attempts. "+
+			"Current block height is %d, transaction was broadcast before finalized block %d. This transaction may not have not been sent and will be marked as fatally errored. "+
+			"This can happen if there is another instance of chainlink running that is using the same private key, or if "+
+			"an external wallet has been used to send a transaction from account %s with nonce %s."+
+			" Please note that Chainlink requires exclusive ownership of it's private keys and sharing keys across multiple"+
+			" chainlink instances, or using the chainlink keys with an external wallet is NOT SUPPORTED and WILL lead to missed transactions",
+			oldTx.ID, head.BlockNumber(), latestFinalizedHead.BlockNumber(), oldTx.FromAddress, oldTx.Sequence.String()), "txID", oldTx.ID, "sequence", oldTx.Sequence.String(), "fromAddress", oldTx.FromAddress)
+
+		// Signal pending tasks for these transactions as failed
+		// Store errors and continue to allow all transactions a chance to be signaled
+		if f.resumeCallback != nil && oldTx.PipelineTaskRunID.Valid && oldTx.SignalCallback && !oldTx.CallbackCompleted {
+			err = f.resumeCallback(ctx, oldTx.PipelineTaskRunID.UUID, nil, errors.New(ErrCouldNotGetReceipt))
+			if errors.Is(err, sql.ErrNoRows) {
+				f.lggr.Debugw("callback missing or already resumed", "etxID", oldTx.ID)
+			} else if err != nil {
+				errorList = append(errorList, fmt.Errorf("failed to resume pipeline for ID %s: %w", oldTx.PipelineTaskRunID.UUID.String(), err))
+			} else {
+				// Mark tx as having completed callback
+				if err = f.txStore.UpdateTxCallbackCompleted(ctx, oldTx.PipelineTaskRunID.UUID, f.chainID); err != nil {
+					errorList = append(errorList, fmt.Errorf("failed to update callback as complete for tx ID %d: %w", oldTx.ID, err))
+				}
+			}
+		}
+	}
+	if len(errorList) > 0 {
+		return errors.Join(errorList...)
+	}
+
+	// Mark transactions as fatal error
+	err = f.txStore.UpdateTxFatalError(ctx, oldTxIDs, ErrCouldNotGetReceipt)
```

Comment:
Shouldn't we call `UpdateTxFatalErrorAndDeleteAttempts` instead?

Reply:
By calling `UpdateTxFatalError` instead, I wanted to preserve some history that this tx had attempts and never made it on-chain. Otherwise, `UpdateTxFatalErrorAndDeleteAttempts` also clears other fields like nonce which I think we want to preserve so our local state shows that nonce as being used. Neither of these might really matter but seemed better to hold on to info rather than clear it here.

Reply:
I thought the only distinction between overflowed and regular fatal transactions was the fact that the first ones don't have attempts, otherwise, we would retry to fetch receipts for them.

Reply:
Oh right good point, just updated the logic to use `UpdateTxFatalErrorAndDeleteAttempts`